### PR TITLE
Initial setup for terminals and icons

### PR DESCRIPTION
**Pull request by PeterMeisrimelModelon**

### CODE REVIEW

File: Test/FMI1/fmi1_import_type_definitions_test.c
Comment (PeterMeisrimelModelon) [inline Test/FMI1/fmi1_import_type_definitions_test.c:32]: unused.

File: src/XML/src/FMI3/fmi3_xml_parser.c
Comment (PeterMeisrimelModelon) [inline src/XML/src/FMI3/fmi3_xml_parser.c:277]: These changes were requires to allow freeing the parser context without everything fully setup/allocated, as is the case when terminalsAndIcons does not exist.

File: src/XML/src/FMI3/fmi3_xml_parser.h
Comment (PeterMeisrimelModelon) [inline src/XML/src/FMI3/fmi3_xml_parser.h:182]: The new elements could have their own list, but won't fix this for this ticket.

File: src/Import/include/FMI/fmi_import_util.h
Comment (modelonrobinandersson) [inline src/Import/include/FMI/fmi_import_util.h:56]:
```suggestion
/** Given directory name fmu_unzipped_path and the model identifier - construct the XML file name

```

File: src/Import/include/FMI3/fmi3_import_terminals_and_icons.h
Comment (modelonrobinandersson) [inline src/Import/include/FMI3/fmi3_import_terminals_and_icons.h:59]:
```suggestion
    @return terminal pointer. NULL if terminal with given name does not exist.
```

File: src/Import/src/FMI3/fmi3_import.c
Comment (modelonrobinandersson) [inline src/Import/src/FMI3/fmi3_import.c:70]: What does this do and/or why is it necessary? Perhaps we can add a comment above to explain it.

File: src/Import/src/FMI3/fmi3_import.c
Comment (modelonrobinandersson) [inline src/Import/src/FMI3/fmi3_import.c:132]:
```suggestion
   // Only parse terminals and icons if parsing of model description did not fail
```

File: src/Import/src/FMI3/fmi3_import.c
Comment (modelonrobinandersson) [inline src/Import/src/FMI3/fmi3_import.c:137]: Curious: Does it matter if we set fmu to `0` or `NULL`? I am only asking this out of consistency as it seems we set to `NULL` usually when there are errors.

File: src/XML/include/FMI3/fmi3_xml_terminals_and_icons.h
Comment (modelonrobinandersson) [inline src/XML/include/FMI3/fmi3_xml_terminals_and_icons.h:41]: Would be good with a comment that explains the purpose of these.

File: src/XML/src/FMI3/fmi3_xml_terminals_and_icons.c
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_terminals_and_icons.c:73]: What does this do?

File: src/XML/src/FMI3/fmi3_xml_terminals_and_icons.c
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_terminals_and_icons.c:151]:
```suggestion
                            "Two terminals with the same name '%s' found. This is not allowed by the specification.", name1);
```
Consider this optional.

File: src/XML/src/FMI3/fmi3_xml_terminals_and_icons_impl.h
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_terminals_and_icons_impl.h:44]: I would suggest adding a TODO comment here that we currently only support `name`

**Review by modelonrobinandersson — APPROVED** [review]
Well done, I only had minor nitpick comments that I don't consider in need of a second review.

File: src/Import/src/FMI3/fmi3_import.c
Comment (PeterMeisrimelModelon) [reply src/Import/src/FMI3/fmi3_import.c:70]: I think the need to initialize buffers is straight forward, so I added comments to the `fmi3_import_t `struct that should clarify the difference between `logMessageBufferCoded `and `logMessageBufferExpanded`

File: src/Import/src/FMI3/fmi3_import.c
Comment (PeterMeisrimelModelon) [reply src/Import/src/FMI3/fmi3_import.c:137]: from stddef.h:

```
#ifndef __cplusplus
#define NULL ((void *)0)
#else   /* C++ */
#define NULL 0
#endif  /* C++ */
```

So, no real practical difference, but I think it increases readability. And, yet, we are not very consistent with using it.

File: src/XML/src/FMI3/fmi3_xml_terminals_and_icons.c
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_terminals_and_icons.c:73]: cb_free is a function pointer to a function with void return and const char* input.  
We assign it the callbacks->free function, with some extra typecasts, since callbacks->free is defined for jm_voidp = void* inputs.

Also added a comment for this.

File: src/XML/src/FMI3/fmi3_xml_terminals_and_icons_impl.h
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_terminals_and_icons_impl.h:44]: Added.

File: src/Import/src/FMI3/fmi3_import.c
Comment (filip-stenstrom) [reply src/Import/src/FMI3/fmi3_import.c]: Please use NULL everywhere it is a pointer. I've tried to change it at all places I've come across it. It makes it very clear that the type is a pointer and not an int.

(I've not looked at other comments/code here, just stumbled on this thread.)

**Pull request merged**
